module/zfs/metaslab.c

```diff
@@ -1659,9 +1654,92 @@ metaslab_block_picker(range_tree_t *rt, uint64_t *cursor, uint64_t size,
 	*cursor = 0;
 	return (-1ULL);
 }
-#endif /* WITH_DF/CF_BLOCK_ALLOCATOR */
 
-#if defined(WITH_DF_BLOCK_ALLOCATOR)
+typedef enum spa_allocator {
+	SPA_DF_ALLOCATOR = 0,	/* Dynamic Fit (df) block allocator */
+	SPA_CF_ALLOCATOR,	/* Cursor fit block allocator */
+	SPA_NDF_ALLOCATOR,	/* New dynamic fit allocator */
+	SPA_NUM_ALLOCATORS
+} spa_allocator_t;
+
+static uint64_t metaslab_df_alloc(metaslab_t *msp, uint64_t size);
+static uint64_t metaslab_cf_alloc(metaslab_t *msp, uint64_t size);
+static uint64_t metaslab_ndf_alloc(metaslab_t *msp, uint64_t size);
+metaslab_ops_t *metaslab_allocator(spa_t *spa);
+
+static metaslab_ops_t metaslab_allocators[SPA_NUM_ALLOCATORS] = {
+	{ "dynamic", metaslab_df_alloc },
+	{ "cursor", metaslab_cf_alloc },
+	{ "new-dynamic", metaslab_ndf_alloc },
+};
+
+static int
+spa_find_allocator_byname(const char *val)
+{
+	for (int a = 0; a < SPA_NUM_ALLOCATORS; a++) {
+
+		/* todo: for removal when ndf is working */
+		if (a == SPA_NDF_ALLOCATOR) continue;
+
+		if (strcmp(val, metaslab_allocators[a].msop_name) == 0)
+			return (a);
+	}
+	return (-1);
+}
+
+void
+spa_set_allocator(spa_t *spa, const char *allocator)
+{
+	int a = spa_find_allocator_byname(allocator);
+	if (a < 0) a = SPA_DF_ALLOCATOR;
+	spa->spa_active_allocator = a;
+	zfs_dbgmsg("spa allocator: %s\n", metaslab_allocators[a].msop_name);
+}
+
+uint64_t
+spa_get_active_allocator(spa_t *spa)
+{
+	return (spa->spa_active_allocator);
+}
+
+#if defined(_KERNEL)
+int
+param_set_active_allocator_common(const char *val)
+{
+	spa_t *spa = NULL;
+	char *p;
+
+	if (val == NULL)
+		return (SET_ERROR(EINVAL));
+
+	if ((p = strchr(val, '\n')) != NULL)
+		*p = '\0';
+
+	int a = spa_find_allocator_byname(val);
+	if (a < 0)
+		return (SET_ERROR(EINVAL));
+
+	zfs_active_allocator = metaslab_allocators[a].msop_name;
+	zfs_dbgmsg("zfs_active_allocator: %s\n", zfs_active_allocator);
+
+	if (spa_mode_global != SPA_MODE_UNINIT) {
+		mutex_enter(&spa_namespace_lock);
+		while ((spa = spa_next(spa)) != NULL)
+			spa->spa_active_allocator = a;
+		mutex_exit(&spa_namespace_lock);
```

Comment:
I don't see a point to update spa_active_allocator for all the pools.  It will not change anything since metaslab_allocator() is called only once and will not be called again.  Changing allocators live would probably require some live re-initialization, that may be not that easy.

Reply:
removed, thanks